Librispeech Deepspeech JAX workload

sourabh2k15 opened the pull request:
this is the final version that doesn't hang 
this version has grad_clip logic commented as that causes workload to hang on GPUs , we'll fix it in a future PR for e2e run

znado commented on algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py:
```diff
@@ -0,0 +1,732 @@
+    if reverse:
+      inputs = flip_sequences(inputs, lengths)
+
+    recurrent_dropout_mask = None
```
if this is always `None`, can we fully delete it?

znado replied:
I still think we should delete this, but we can leave it for a future cleanup

znado commented on algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py:
```diff
@@ -0,0 +1,732 @@
+class BatchRNN(nn.Module):
+  """Implements a single conformer encoder layer.
+
+  High level overview:
```
is this missing some docstring, or should it be deleted?

znado commented on submission_runner.py:
```diff
@@ -263,13 +262,6 @@ def train_once(
             rng=update_rng)
     except spec.TrainingCompleteError:
       training_complete = True
-    except RuntimeError as e:
-      if "out of memory" in str(e):
-        logging.warning(
-            f'error: GPU out of memory during training during step {global_step}, error: {str(e)}'  # pylint: disable=line-too-long
-        )
-        if torch.cuda.is_available():
-          torch.cuda.empty_cache()
```
should we keep this?

sourabh2k15 replied:
it's extra code that doesn't really catch OOM errors, I included it last time in hope that it might catch a real GPU OOM error

znado commented on algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/workload.py:
```diff
@@ -0,0 +1,50 @@
+
+from algorithmic_efficiency import spec
+from algorithmic_efficiency.workloads.librispeech_conformer.librispeech_jax.workload import \
+    LibriSpeechConformerWorkload
```
do we need to check if model_state is not none here? https://github.com/mlcommons/algorithmic-efficiency/blob/main/algorithmic_efficiency/workloads/librispeech_conformer/librispeech_jax/workload.py#L61

also do we need to return model_state here instead of None? https://github.com/mlcommons/algorithmic-efficiency/blob/main/algorithmic_efficiency/workloads/librispeech_conformer/librispeech_jax/workload.py#L82

sourabh2k15 replied:
1) it will fail automatically if batch_stats is not passed in 
2) we return model_state as None for eval mode and we ignore that returned model_state in eval code path so it's fine

znado commented on reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py:
```diff
@@ -0,0 +1,207 @@
+  opt_init_fn, opt_update_fn = optimizer(hyperparameters,
+                                         workload.num_train_examples)
+  optimizer_state = opt_init_fn(params_zeros_like)
+  # log_pytree_shape_and_statistics(
+  #     flax.jax_utils.unreplicate(model_params), workload.summary_writer)
```
delete

znado commented on reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py:
```diff
@@ -0,0 +1,207 @@
+  return weight_l2
+
+
+def _summary_str(param):
+  total_norm = jnp.linalg.norm(param.reshape(-1))
+  return '{} - {} - {}'.format(str(param.shape), param.size, total_norm)
+
+
+def log_pytree_shape_and_statistics(pytree, writer):
+  """Logs the shape and norm of every array in the pytree."""
+  if not pytree:
+    logging.info('Empty pytree')
+    return
+  logging.info('Printing model param shapes.')
+  shape_dict = jax.tree_map(_summary_str, pytree)
+
+  logging.info(shape_dict.pretty_repr())
+
+  total_params = jax.tree_util.tree_reduce(
+      operator.add, jax.tree_map(lambda x: x.size, pytree))
+  print('num params = ', total_params)
```
delete

sourabh2k15 replied:
done

znado commented on reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py:
```diff
@@ -0,0 +1,207 @@
+  grad_scaling_factor = grad_clip / (grad_norm + _GRAD_CLIP_EPS)
+  grad_scaling_factor = jax.lax.clamp(min=0.0, x=grad_scaling_factor, max=1.0)
+
+  grad = jax.tree_map(lambda x: x * grad_scaling_factor, grad)
+  loss, grad = lax.pmean((loss, grad), axis_name='batch')
```
I don't remember what we settled on, but is it possible to pmean before calculating grad_norm? if not, can you add a comment saying we need pmean after otherwise the code hangs?

sourabh2k15 replied:
so with batch size 256 doesn't matter if grad clip code is before / after pmean it still hangs 
I've commented out the grad_clip code and I'm unsure if this workload even needs gradient clipping , we never ran it without clip 

for now I've kept commented code and we can figure out if we need it / how to make it not hang if we do need it

sourabh2k15 replied:
resolved in newest version of code, pmean happens before grad clipping now 
I've still eliminated jax.lax.cond in our grad clipping code as that's more efficient

znado commented on reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py:
```diff
@@ -0,0 +1,207 @@
+    model_state,
+    optimizer_state)
+
+  print('{}) loss = {}, grad_norm = {}'.format(global_step,
+                                               loss.mean(),
+                                               grad_norm.mean()))
```
delete

sourabh2k15 replied:
converted to logging.info() 
we need this to have visibility into the run

sourabh2k15 commented on algorithmic_efficiency/workloads/librispeech_conformer/workload.py:
```diff
@@ -37,7 +37,7 @@ def num_train_examples(self):
   @property
   def num_eval_train_examples(self):
-    return 512
+    return 512 
```
done

sourabh2k15 commented on algorithmic_efficiency/workloads/librispeech_conformer/workload.py:
```diff
@@ -157,7 +167,7 @@ def _build_dataset(self,
 
     if split == 'train':
-      split = 'train-clean-100+train-clean-360+train-other-500'
+      split = 'train-clean-100'
```
done

sourabh2k15 commented on algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py:
```diff
@@ -0,0 +1,732 @@
+  bidirectional: bool = True
+
+
+class Subsample(nn.Module):
```
okay for now I'll leave it unmerged so whoever's debugging slowness in deepspeech can only look at this file for model code

sourabh2k15 commented on algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py:
```diff
@@ -0,0 +1,732 @@
+    return outputs, out_padding
+
+
+class FeedForwardModule(nn.Module):
```
not touching deepspeech model code for now , will update later

sourabh2k15 commented on algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py:
```diff
@@ -0,0 +1,732 @@
+    return inputs
+
+
+class LayerNorm(nn.Module):
```
will merge common parts between conformer and deepspeech in a future PR

sourabh2k15 commented on algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/e2e_test.py:
```diff
@@ -0,0 +1,251 @@
+from algorithmic_efficiency.workloads.librispeech_deepspeech.librispeech_jax import \
```
moved to tests/ 
I still want to keep it checked in for whoever debugs this workload slowness

sourabh2k15 replied:
removed it for now

sourabh2k15 commented on algorithmic_efficiency/workloads/librispeech_conformer/input_pipeline.py:
```diff
@@ -6,12 +6,11 @@
 def get_librispeech_dataset(split_name: str,
                             data_dir: str,
-                            shuffle_rng: spec.RandomState,
+                            shuffle_rng,
```
done

sourabh2k15 commented on algorithmic_efficiency/workloads/librispeech_conformer/input_pipeline.py:
```diff
@@ -6,12 +6,11 @@
 
-from algorithmic_efficiency import spec
-
+tf.config.set_visible_devices([], 'GPU')
```
removed

znado commented on reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py:
```diff
@@ -0,0 +1,187 @@
+      rng=rng,
+      lr=lr)
+
+  pmapped_update_step = jax.pmap(
+      update_fn, axis_name='batch', in_axes=(0, 0, 0, 0))
```
we need to refactor this to avoid pmap inside the loop like in conformer, right? https://github.com/mlcommons/algorithmic-efficiency/blob/main/reference_submissions/librispeech_conformer/librispeech_jax/submission.py#L80

sourabh2k15 replied:
that's correct, updated pmap code to be same as conformer submission.py

znado commented on setup.cfg:
```diff
@@ -36,7 +36,7 @@ setup_requires =
 install_requires =
   absl-py==1.0.0
-  numpy==1.21.6
+  numpy==1.23.3
```
we need to undo this change, I think it'll break on lower versions of python

znado commented on tests/e2e_test.py:
```diff
@@ -0,0 +1,230 @@
+# pylint: skip-file
```
I think we should hold off on checking this in for now (but definitely still keep it around in your fork!!).

at some point we'll need to write up serious e2e testing for _all_ workloads, so we'll have this functionality then anyways

znado replied:
(also removing this should fix the remaining lint issue)

sourabh2k15 replied:
sounds good, deleted

znado commented on algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py:
```diff
@@ -0,0 +1,730 @@
+    An ndarray with the flipped inputs.
+  """
+  # Compute the indices to put the inputs in flipped order as per above example.
+  max_length = inputs.shape[0]
```
instead of taking `lengths` as an input, I think we can recompute it as `lengths = jnp.argmin(inputs, axis=1)`

sourabh2k15 replied:
made the change to pass input_paddings as deep as possible

znado commented on algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py:
```diff
@@ -0,0 +1,730 @@
+                                                 inputs,
+                                                 recurrent_dropout_mask,
+                                                 deterministic)
+    final_state = jax.tree_map(
+        lambda x: x[jnp.arange(inputs.shape[0]), lengths - 1], cell_states)
```
as far as I can tell, we never use this, so lets delete this code

sourabh2k15 replied:
deleted

znado commented on algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py:
```diff
@@ -0,0 +1,730 @@
+                       config.batch_norm_epsilon)(inputs, input_paddings, train)
+    lengths = jnp.sum(1 - input_paddings, axis=-1, dtype=jnp.int32)
+
+    output, _ = LSTM(
```
return just outputs and delete the `, _`

sourabh2k15 replied:
deleted

znado commented on algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py:
```diff
@@ -0,0 +1,730 @@
+      for others cells it only contains a single vector (h,).
+    """
+    batch_size = inputs.shape[0]
+    final_states = []
```
we never use `final_states`, so we should be able to delete this

sourabh2k15 replied:
deleted final_states

znado commented on reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py:
```diff
@@ -0,0 +1,187 @@
+
+from algorithmic_efficiency import spec
+
+# import gc
+# import torch
+
```
delete?

sourabh2k15 replied:
deleted

znado commented on reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py:
```diff
@@ -0,0 +1,187 @@
+  grad_fn = jax.value_and_grad(_loss_fn, has_aux=True)
+  (loss, new_model_state), grad = grad_fn(params)
+
+  grad_norm = jnp.sqrt(l2_regularization(grad, 0))
```
(we can follow up offline too, but)

is it possible to flip the order of these lines?

```
  loss, grad = lax.pmean((loss, grad), axis_name='batch')
  grad_norm = jnp.sqrt(l2_regularization(grad, 0))
```

sourabh2k15 replied:
fixed it in latest deepspeech submission code, now pmean happens before grad clipping

sourabh2k15 commented on algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py:
```diff
@@ -0,0 +1,730 @@
+  return jnp.where(x < min_value,
+                   min_value,
+                   jnp.where(x > max_value, max_value, x))
+
```
removed

sourabh2k15 commented:
update : this PR is good to go, here's the e2e run : https://tensorboard.dev/experiment/45h1nSXfRqypQdgpN6Dgyw 

step time looks good, will check tomorrow if it achieves expected target validation WER 

changes done to model in final state :
1) removed a BatchNorm in convolution subsampling layer which is not present in conformer and is unclear if present in original paper , this specific BatchNorm causes a 2 minute constant folding compilation stage on GPU causing slowdown 
2) fixed a bug in  FeedForwardModule , we were missing relu activation function 
3) add cross device syncing code in BatchNorm so mean and variance are computed from the entire batch not just sharded batch 
4) put grad_clip = 5.0 given initial gradient norms do go higher from tensorboard 

changes done to submission code to resolve recompilation : 
pmapped train step was written in a way that triggers recompilation at every step, this was resolved by adopting the pmap code from conformer submission file 

debugging note : can use JAX_LOG_COMPILES=1 to see whenever jax recompiles stuff

znado reviewed (approved):
Looks great, it's fantastic to finally get this in!
